src/app/clusters/power-source-server/power-source-server.cpp

```diff
@@ -55,20 +105,111 @@ CHIP_ERROR PowerSourceAttrAccess::Read(const ConcreteReadAttributePath & aPath,
         // TODO: Needs implementation.
         err = aEncoder.EncodeEmptyList();
         break;
-    case EndpointList::Id:
-        // TODO: Needs implementation and a way to allow dynamic endpoints to register endpoints
-        err = aEncoder.EncodeEmptyList();
+    case EndpointList::Id: {
+        PowerSourceServer & server    = PowerSourceServer::Instance();
+        const Span<EndpointId> * span = server.GetEndpointList(aPath.mEndpointId);
+        if (span == nullptr)
+        {
+            err = aEncoder.EncodeEmptyList();
+        }
+        else
+        {
+            DataModel::List<EndpointId> list(*span);
+            err = aEncoder.Encode(list);
+        }
         break;
+    }
     default:
         break;
     }
 
     return err;
 }
 
-} // anonymous namespace
+PowerSourceAttrAccess & TestOnlyGetPowerSourceAttrAccess()
+{
+    return gAttrAccess;
+}
 
-void MatterPowerSourcePluginServerInitCallback()
+PowerSourceServer & PowerSourceServer::Instance()
 {
-    registerAttributeAccessOverride(&gAttrAccess);
+    return gPowerSourceServer;
+}
+
+// Caller does not need to retain the span past the call point as these are copied into an internal storage
+CHIP_ERROR PowerSourceServer::SetEndpointList(EndpointId powerSourceClusterEndpoint, Span<EndpointId> endpointList)
+{
+    // TODO: should check here that the power source cluster exists on the endpoint, but for now let's take the caller's word for it
+
+    size_t idx = PowerSourceClusterEndpointIndex(powerSourceClusterEndpoint);
+    if (idx >= kNumSupportedEndpoints)
+    {
+        idx = NextEmptyIndex();
+    }
+    if (idx >= kNumSupportedEndpoints)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    sPowerSourceClusterInfo[idx].Clear();
+    if (endpointList.size() == 0)
+    {
+        sPowerSourceClusterInfo[idx] = PowerSourceClusterInfo();
+    }
+    else
+    {
+        sPowerSourceClusterInfo[idx] = PowerSourceClusterInfo(powerSourceClusterEndpoint);
+        sPowerSourceClusterInfo[idx].SetEndpointList(endpointList);
+    }
+    return CHIP_NO_ERROR;
+}
+const Span<EndpointId> * PowerSourceServer::GetEndpointList(EndpointId powerSourceClusterEndpoint) const
+{
+    size_t idx = PowerSourceClusterEndpointIndex(powerSourceClusterEndpoint);
+    if (idx != std::numeric_limits<size_t>::max())
+    {
+        return &sPowerSourceClusterInfo[idx].mEndpointList;
+    }
+    return nullptr;
+}
+
+void PowerSourceServer::Shutdown()
+{
+    for (size_t i = 0; i < kNumSupportedEndpoints; ++i)
+    {
+        sPowerSourceClusterInfo[i].Clear();
+    }
+}
+
+size_t PowerSourceServer::GetNumSupportedEndpointLists() const
+{
+    return kNumSupportedEndpoints;
 }
+
+size_t PowerSourceServer::PowerSourceClusterEndpointIndex(EndpointId endpointId) const
```

Comment:
This is probably OK, but we have existing utilities (`emberAfGetClusterServerEndpointIndex`) for things like "map endpoint id to a list index, taking into account the fixed endpoints that we know have this cluster and the fact that all dynamic endpoints might have it"...  Is there a reason we're not just using that?

Reply:
The endpoint count is different.

Reply:
@cecille Why is it different, exactly?